resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala

```diff
@@ -73,7 +73,8 @@ private[spark] class KubernetesDriverBuilder {
       new HadoopConfDriverFeatureStep(conf),
       new KerberosConfDriverFeatureStep(conf),
       new PodTemplateConfigMapStep(conf),
-      new LocalDirsFeatureStep(conf)) ++ userFeatures
+      new LocalDirsFeatureStep(conf),
+      new DriverIngressFeatureStep(conf)) ++ userFeatures
```

Comment:
Can we provide a way for Spark Job runs successfully even `DriverIngressFeatureStep` and its K8s resources fails? IIUC, if something goes wrong, the Spark job is not started, isn't it?

Reply:
that's a good point. 

yes, any failure during resource creation will cause the job submission failure at the current stage, and I did see the ingress creation failure cases internally. we mitigate it by providing a switch to allow users to disable creating ingress for certain jobs `spark.kubernetes.driver.ingress.enabled`.

currently, we create resources in batches,

https://github.com/apache/spark/blob/48e207f4a2192d474f2c0f141b984ef0c36a78c3/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala#L161-L181

if we want to make some resources optional, we need to add a new batch to handle them independently. meanwhile, a new method may need to add in `KubernetesFeatureConfigStep`, something like

```
trait KubernetesFeatureConfigStep {
  ...
  def getAdditionalOptionalKubernetesResources(): Seq[Option[HasMetadata]] = Seq.empty
}
```

@dongjoon-hyun WDYT?

Reply:
we need a decision here.

option 1: provide a switch `spark.kubernetes.driver.ingress.enabled` to allow the user to explicitly enable/disable ingress, and fail the job when creating ingress failed.

pros:
1. behavior consistent with `spark.ui.enabled`;
2. if ingress creation failed, the user knows what happened rather than confusing why Spark UI is unavailable;
3. code change is simple;

cons:
1. ingress is actually an optional resource, technically, Spark could run well without ingress.

option 2: treat ingress as an optional resource, ignore ingress creation failure
pros:
1. Keep the same stability for Spark jobs submission even with ingress enabled.

cons:
1. ingress resource is not reliable even with explicitly enabling
2. we need to touch the developer interface to achieve such functionalities (there is one proposal above)

internally, we adopted option 1, and some users explicitly disabled the ingress because they didn't want the job submission to fail because of ingress creation failure.

@dongjoon-hyun @yaooqinn @mridulm @LuciferYang please let me know which one are you preferred.

Reply:
To @pan3793 , I'm very careful about this PR because we consider `Having K8s Ingress` is more vulnerable than having `Spark UI` because it increases the boundary of exposure. `Spark UI` itself is not open to the other pods if you design to isolate Spark pod-to-pod communication carefully.

In addition, some K8s implementations (and their security backend) increases `K8s Ingress` creation. I expected a few minutes delays for only K8s ingress entity creations. It could mean a delay of Spark Job time too if we design to wait for K8s ingress.

Since we verified that the deletion is handled by K8s ownership correctly. Let's double-check the following.
- What is the job start time for Option 1? There is no delay?
  - If there is a delay, is there a way to mitigate it?
  - If there is no delay because it is handled separately, what happens during the time gap after the job start already and `K8s Ingress` is still under construction. Are we in the same situation where `Option 2`'s `Cons 1`?

Reply:
> I expected a few minutes delays for only K8s ingress entity creations. 

that surprises me, in our cluster, the overhead of ingress creation is negligible.

I tested with ingress enabled/disabled 5 times in internal cluster to measure the job submission time by `spark.app.startTime` - `spark.app.submitTime`.

ingress enabled:  20715, 24954, 17129, 16631, 16076
ingress disabled: 20246, 20262, 17176, 18633, 19962

the difference is majorly caused by Pod creation, ingress creation is negligible.

if the ingress creation is too heavy, option 2 is the only choice.